compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/AwtDragAndDropManager.desktop.kt

```diff
@@ -174,140 +182,193 @@ internal class AwtDragAndDropManager(
-                val event = DragAndDropEvent(dte)
-                rootDragAndDropNode.onExited(event)
-                endDrag(event)
-            }
+    override fun getSourceActions(c: JComponent?): Int {
+        val actions = outgoingTransfer?.transferData?.supportedActions ?: emptyList()
+        return actions.fold(
+            initial = NONE,
+            operation = { acc, action -> acc or action.awtAction },
+        )
+    }
 
-            override fun dragOver(dtde: DropTargetDragEvent) {
-                rootDragAndDropNode.onMoved(DragAndDropEvent(dtde))
-            }
+    override fun getDragImage() = outgoingTransfer?.dragImage
 
-            override fun dropActionChanged(dtde: DropTargetDragEvent) {
-                rootDragAndDropNode.onChanged(DragAndDropEvent(dtde))
-            }
+    override fun getDragImageOffset() = outgoingTransfer?.dragImageOffset
 
-            override fun drop(dtde: DropTargetDropEvent) {
-                val event = DragAndDropEvent(dtde)
-                dtde.acceptDrop(dtde.dropAction)
-                dtde.dropComplete(rootDragAndDropNode.onDrop(event))
-                endDrag(event)
-            }
+    override fun exportDone(source: JComponent?, data: Transferable?, action: Int) {
+        super.exportDone(source, data, action)
 
-            private fun endDrag(event: DragAndDropEvent) {
-                rootDragAndDropNode.onEnded(event)
-                interestedNodes.clear()
-            }
+        val transferAction = DragAndDropTransferAction.fromAwtAction(action)
+        outgoingTransfer?.transferData?.onTransferCompleted?.invoke(transferAction)
+        outgoingTransfer = null
+    }
+}
 
-            private fun DragAndDropEvent(dragEvent: DropTargetDragEvent) = DragAndDropEvent(
-                nativeEvent = dragEvent,
-                action = DragAndDropTransferAction.fromAwtAction(dragEvent.dropAction),
-                positionInRootImpl = dragEvent.location.toOffset()
-            )
+private class OutgoingTransfer(
+    val transferData: DragAndDropTransferData,
+    val dragImage: Image,
+    val dragImageOffset: Point
+)
 
-            private fun DragAndDropEvent(dropEvent: DropTargetDropEvent) = DragAndDropEvent(
-                nativeEvent = dropEvent,
-                action = DragAndDropTransferAction.fromAwtAction(dropEvent.dropAction),
-                positionInRootImpl = dropEvent.location.toOffset()
-            )
+/**
+ * The AWT [DropTarget] we install in the root container in order to implement drop-target
+ * functionality.
+ *
+ * Note that one instance may be shared between several [AwtDragAndDropManager]s, if there are
+ * several compositions attached to the same root container. It forwards each [DropTargetEvent] it
+ * receives to the appropriate [AwtDragAndDropManager.dropTargetListener].
+ */
+private class ComposeDropTarget(
+    rootContainer: JComponent,
+    private val activeDragAndDropManager: () -> PlatformDragAndDropManager
```

Comment:
It looks like it requires only `dropTargetListener`, not entire `DropTargetListener`

Reply:
It does, but this is local to the implementation (not public API), so it's not a big deal, and it avoids creating an additional lambda like this:
```
ComposeDropTarget(
    activeDropListener = { activeDragAndDropManager().dropTarget }
)
```